lib/Sema/TypeCheckMacros.cpp

```diff
@@ -214,8 +214,9 @@ Expr *swift::expandMacroExpr(
 
   // Create a new source buffer with the contents of the expanded macro.
   auto macroBuffer =
-      llvm::MemoryBuffer::getMemBuffer(evaluatedSource, bufferName);
+      llvm::MemoryBuffer::getMemBufferCopy(evaluatedSource, bufferName);
   unsigned macroBufferID = sourceMgr.addNewSourceBuffer(std::move(macroBuffer));
+  free((void*)evaluatedSource.data());
```

Comment:
We know it works. But still a little scary.

It would be great if there's a way to pass C++ lambda to `swift_ASTGen_evaluateMacro` where
* the lambda receives a temporary c++ string (i.e. handler callback)
    ```cpp
    swift_ASTGen_evaluateMacro( ..., [&](unsigned char *utf8Start, size_t length) {
      // utf8Start is valid only while this lambda.
      macroBuffer = llvm::MemoryBuffer::getMemBufferCopy({utf8Start, length});
    });
    ```
* or, the lambda receives a size and returns a buffer pointer of the size (i.e. factory)
    ```cpp
    swift_ASTGen_evaluateMacro( ..., [&](size_t utf8Length) {
      macroBuffer = llvm::WritableMemoryBuffer::getNewUninitMemBuffer(utf8Length + 1, bufferIdentifier);
      return macroBuffer->getBufferStart();
      // the caller initialize the buffer.
    });
    ```
So the size that allocates the memory has the responsibility to free it.

But I guess there's no way to pass C++ lambda to Swift?

Reply:
We can pass through a C function pointer and a context pointer, which would eliminate a copy and not have a malloc/free pair across languages.